Add rules for dense matrix exponential

sethaxen opened the pull request:
Fixes #331

Because I plan to add a number of matrix functions, and they are long, I have added these rules to a new function `matfun.jl`.
These rules unfortunately require quite a bit of code duplication from LinearAlgebra, but following discussion in https://github.com/JuliaLang/julia/issues/5840, that's only unavoidable if we refactor these functions in LinearAlgebra to return their intermediates.

sethaxen commented on src/rulesets/LinearAlgebra/matfun.jl:
```diff
@@ -0,0 +1,243 @@
+# NOTE: for matrix functions whose power series representation has real coefficients,
+# the pullback and pushforward are related by an adjoint.
+# Specifically, if the pushforward of f(A) is (f_*)_A(ΔA), then the pullback at Y=f(A) is
+# (f^*)_Y(ΔY) = (f_*)_{A'}(ΔY) = ((f_*)_A(ΔY'))'
```
Well that is hideous, but notation is hard, and harder in unicode.

sethaxen replied:
>What if we just made a section for this in the docs, (maybe as internal notes or something)
> and wrote the latex and then linked to that?

Hm, that's an idea. I'll consider it, potentially for a future PR.

> This bit makes sense:
> `(f^*)_Y(ΔY) = ((f_*)_A(ΔY'))'`
> so the pullback at `A`, i.e. the pullback from Y...is equal to the the adjoint of the pushing forward at `A`, the adjoint of of the output senstivity.

Ah yes your description is correct (although it's the adjoint of the pushing forward of the adjoint). I just checked Lee, and this should be the right notation:
```suggestion
# (f^*)_A(ΔY) = (f_*)_{A'}(ΔY) = ((f_*)_A(ΔY'))'
```

> (though that's not well defined since not all functions are monotonic?)

I'm not sure what you mean by this.

> Magical expodential symmetry? (I feel like i made the same suprised sounds for the same reason on your last PR)

It's still surprising to me. Although this property is general for all of the matrix functions defined in LinearAlgebra, not just `exp`. It doesn't follow for all matrix functions though, just those whose convergent power series have *real* coefficients.

sethaxen commented on src/rulesets/LinearAlgebra/matfun.jl:
```diff
@@ -0,0 +1,243 @@
+"""
+    _matfun_frechet!(f, A, Y, ΔA, intermediates)
+
+Similar to `_matfun_frechet!`, but where `ΔA` may be overwritten.
```
```suggestion
Similar to [`_matfun_frechet`](@ref), but where `ΔA` may be overwritten.
```

oxinabox commented on src/rulesets/LinearAlgebra/matfun.jl:
```diff
@@ -0,0 +1,243 @@
+## Adapted from LinearAlgebra.exp! with return of intermediates
+function _matfun!(::typeof(exp), A::StridedMatrix{T}) where T<:BlasFloat
+    n = LinearAlgebra.checksquare(A)
+    ilo, ihi, scale = LAPACK.gebal!('B', A)    # modifies A
+    nA   = opnorm(A, 1)
+    Inn    = Matrix{T}(I, n, n)
```
```suggestion
    ilo, ihi, scale = LAPACK.gebal!('B', A)  # modifies A
    nA = opnorm(A, 1)
    Inn = Matrix{T}(I, n, n)
```

oxinabox commented on src/rulesets/LinearAlgebra/matfun.jl:
```diff
@@ -0,0 +1,243 @@
+    ## For sufficiently small nA, use lower order Padé-Approximations
+    if (nA <= 2.1)
+        if nA > 0.95
+            C = T[17643225600.,8821612800.,2075673600.,302702400.,
+                     30270240.,   2162160.,    110880.,     3960.,
+                           90.,         1.]
+        elseif nA > 0.25
+            C = T[17297280.,8648640.,1995840.,277200.,
+                     25200.,   1512.,     56.,     1.]
+        elseif nA > 0.015
+            C = T[30240.,15120.,3360.,
+                    420.,   30.,   1.]
+        else
+            C = T[120.,60.,12.,1.]
+        end
+        si = 0
+    else
+        C = T[64764752532480000.,32382376266240000.,7771770303897600.,
+                1187353796428800.,  129060195264000.,  10559470521600.,
+                    670442572800.,      33522128640.,      1323241920.,
+                        40840800.,           960960.,           16380.,
+                             182.,                1.]
```
In generalk I am ok with this formatting, except should space after comma.
It did make me think this was a matrix rather than a vector initially.
But it's fine

```suggestion
            C = T[17643225600., 8821612800., 2075673600., 302702400.,
                     30270240.,    2162160.,     110880.,      3960.,
                           90.,          1.]
        elseif nA > 0.25
            C = T[17297280., 8648640., 1995840., 277200.,
                     25200.,    1512.,      56.,      1.]
        elseif nA > 0.015
            C = T[30240., 15120., 3360.,
                    420.,    30.,    1.]
        else
            C = T[120., 60., 12., 1.]
        end
        si = 0
    else
        C = T[64764752532480000., 32382376266240000., 7771770303897600.,
                1187353796428800.,   129060195264000.,   10559470521600.,
                    670442572800.,       33522128640.,       1323241920.,
                        40840800.,            960960.,            16380.,
                             182.,                 1.]
```

nickrobinson251 replied:
fwiw, JuliaFormatter BlueStyle tells me this "should" be
```suggestion
            C = T[
                17643225600.0,
                8821612800.0,
                2075673600.0,
                302702400.0,
                30270240.0,
                2162160.0,
                110880.0,
                3960.0,
                90.0,
                1.0,
            ]
        elseif nA > 0.25
            C = T[17297280.0, 8648640.0, 1995840.0, 277200.0, 25200.0, 1512.0, 56.0, 1.0]
        elseif nA > 0.015
            C = T[30240.0, 15120.0, 3360.0, 420.0, 30.0, 1.0]
        else
            C = T[120.0, 60.0, 12.0, 1.0]
        end
        si = 0
    else
        C = T[
            64764752532480000.0,
            32382376266240000.0,
            7771770303897600.0,
            1187353796428800.0,
            129060195264000.0,
            10559470521600.0,
            670442572800.0,
            33522128640.0,
            1323241920.0,
            40840800.0,
            960960.0,
            16380.0,
            182.0,
            1.0,
        ]
```

nickrobinson251 replied:
i think running JuliaFormatter on this file would be good :)

oxinabox commented on src/rulesets/LinearAlgebra/matfun.jl:
```diff
@@ -0,0 +1,243 @@
+    W  = C[2] * P
+    V  = C[1] * P
+    Apows = typeof(P)[]
+    for k in 1:(div(size(C, 1), 2) - 1)
```
```suggestion
    for k in 1:(size(C, 1) ÷ 2 - 1)
```

oxinabox commented on src/rulesets/LinearAlgebra/matfun.jl:
```diff
@@ -0,0 +1,243 @@
+    end
+
+    if si > 0
+        A ./= convert(T,2^si)
```
```suggestion
        A ./= convert(T, 2^si)
```

oxinabox commented on src/rulesets/LinearAlgebra/matfun.jl:
```diff
@@ -0,0 +1,243 @@
+                        40840800.,           960960.,           16380.,
+                             182.,                1.]
+        s = log2(nA/5.4)               # power of 2 later reversed by squaring
+        si = ceil(Int,s)
```
```suggestion
        si = ceil(Int, s)
```

oxinabox commented on src/rulesets/LinearAlgebra/matfun.jl:
```diff
@@ -0,0 +1,243 @@
+    end
+    U = A * W
+    X = V + U
+    F = lu!(V-U) # NOTE: use lu! instead of LAPACK.gesv! so we can reuse factorization
```
```suggestion
    F = lu!(V - U)  # NOTE: we use lu! instead of LAPACK.gesv! so we can reuse factorization
```

oxinabox commented on src/rulesets/LinearAlgebra/matfun.jl:
```diff
@@ -0,0 +1,243 @@
+    ldiv!(F, X)
+    Xpows = typeof(X)[X]
+    if si > 0            # squaring to reverse dividing by power of 2
+        for t=1:si
```
```suggestion
        for t in 1:si
```

oxinabox commented on src/rulesets/LinearAlgebra/matfun.jl:
```diff
@@ -0,0 +1,243 @@
+    F = lu!(V-U) # NOTE: use lu! instead of LAPACK.gesv! so we can reuse factorization
+    ldiv!(F, X)
+    Xpows = typeof(X)[X]
+    if si > 0            # squaring to reverse dividing by power of 2
```
```suggestion
    if si > 0  # squaring to reverse dividing by power of 2
```

oxinabox commented on src/rulesets/LinearAlgebra/matfun.jl:
```diff
@@ -0,0 +1,243 @@
+    end
+
+    # Undo the balancing
+    for j = ilo:ihi
```
```suggestion
    for j in ilo:ihi
```

oxinabox commented on src/rulesets/LinearAlgebra/matfun.jl:
```diff
@@ -0,0 +1,243 @@
+    # Undo the balancing
+    for j = ilo:ihi
+        scj = scale[j]
+        for i = 1:n
```
```suggestion
        for i in 1:n
```

oxinabox commented on src/rulesets/LinearAlgebra/matfun.jl:
```diff
@@ -0,0 +1,243 @@
+    for j = ilo:ihi
+        scj = scale[j]
+        for i = 1:n
+            X[j,i] *= scj
```
```suggestion
            X[j, i] *= scj
```

oxinabox commented on src/rulesets/LinearAlgebra/matfun.jl:
```diff
@@ -0,0 +1,243 @@
+        for i = 1:n
+            X[j,i] *= scj
+        end
+        for i = 1:n
```
```suggestion
        for i in 1:n
```

oxinabox commented on src/rulesets/LinearAlgebra/matfun.jl:
```diff
@@ -0,0 +1,243 @@
+            X[j,i] *= scj
+        end
+        for i = 1:n
+            X[i,j] /= scj
```
```suggestion
            X[i, j] /= scj
```

oxinabox reviewed (commented):
I started to comment on the formatting, but its a bit much.
I think just run https://github.com/domluna/JuliaFormatter.jl/ over it
`format_file("matfun.jl", BlueStyle())`
its pretty good.

I will review after that, since then I won't be spending time on the basic stuff.

sethaxen commented:
> I started to comment on the formatting, but its a bit much.
Yeah, it's the formatting used by `exp!` in LinearAlgebra, which is not great.

> I think just run https://github.com/domluna/JuliaFormatter.jl/ over it
> `format_file("matfun.jl", BlueStyle())`
Done!

sethaxen commented on src/rulesets/LinearAlgebra/matfun.jl:
```diff
@@ -0,0 +1,261 @@
+        end
+    end
+
+    if ilo > 1  # apply lower permutations in reverse order
```
I haven't been able to find a matrix `A` that hits this branch. As far as I can tell, all `exp` tests in LinearAlgebra also do not hit the corresponding branch, but code coverage indicates it is covered, so it must be indirectly hit by one of the trigonometric/hyperbolic matrix function tests or something.

sethaxen replied:
Nevermind, found one

sethaxen commented on src/rulesets/LinearAlgebra/matfun.jl:
```diff
@@ -0,0 +1,270 @@
+        A = copy(A0)
+        X, intermediates = _matfun!(exp, A)
+        function exp_pullback(ΔX)
+            ΔX′ = copy(adjoint(ΔX))
```
Perhaps. It might make sense to explicitly introduce a `_matfun_frechet_adjoint` and in-place version that handle the adjoint logic, which will be a little different for every matrix type.

sethaxen commented:
@oxinabox I added a `_matfun_frechet_adjoint`, changed the signature to put the differential first, and substantially modified the docstrings and comments. Would you mind re-reviewing the comments and docstrings at the top of `matfun.jl` before I merge?

oxinabox reviewed (approved):
LGTM
